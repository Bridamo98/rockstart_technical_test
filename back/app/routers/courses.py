from typing import List

from app.models.models import Courses
from app.modules.factory import factory_for
from app.modules.schemas import CourseCreate, CourseOut, CourseUpdate
from fastapi import APIRouter, HTTPException

router = APIRouter(
    prefix="/courses",
    tags=["courses"],
)

COURSE_NOT_FOUND_MSG = "Course not found"


@router.post("/", response_model=CourseOut)
def create_course(payload: CourseCreate):
    with factory_for(Courses) as f:
        inst = Courses.create(
            title=payload.title,
            course_desc=payload.course_desc,
            instructor_id=payload.instructor_id,
        )
        obj = f.create(obj=inst, refresh=True)
        return f.serialize(obj)


@router.get("/", response_model=List[CourseOut])
def list_courses():
    with factory_for(Courses) as f:
        objs = f.get_all()
        return f.serialize_many(objs)


@router.get("/{course_id}", response_model=CourseOut)
def get_course(course_id: int):
    with factory_for(Courses) as f:
        obj = f.get_by_id(course_id)
        if not obj:
            raise HTTPException(status_code=404, detail=COURSE_NOT_FOUND_MSG)
        return f.serialize(obj)


@router.put("/{course_id}", response_model=CourseOut)
def update_course(course_id: int, payload: CourseUpdate):
    changes = payload.model_dump(exclude_unset=True)
    with factory_for(Courses) as f:
        obj = f.update_by_id(course_id, refresh=True, **changes)
        if not obj:
            raise HTTPException(status_code=404, detail=COURSE_NOT_FOUND_MSG)
        return f.serialize(obj)


@router.delete("/{course_id}")
def delete_course(course_id: int):
    with factory_for(Courses) as f:
        success = f.delete(course_id)
        if not success:
            raise HTTPException(status_code=404, detail=COURSE_NOT_FOUND_MSG)
        return {"deleted": course_id}
