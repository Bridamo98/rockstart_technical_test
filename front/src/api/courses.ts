import type { Course, CourseCreateDTO, CourseUpdateDTO } from "../types";
import { api } from "./client";

export const CoursesAPI = {
  list: () => api.get<Course[]>("/courses/"),
  get: (id: number) => api.get<Course>(`/courses/${id}`),
  create: (payload: CourseCreateDTO) => api.post<Course>("/courses/", payload),
  update: (id: number, payload: CourseUpdateDTO) =>
    api.put<Course>(`/courses/${id}`, payload),
  remove: (id: number) => api.delete(`/courses/${id}`),
};
