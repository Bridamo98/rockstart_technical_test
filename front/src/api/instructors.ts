import type { Instructor } from "../types";
import { api } from "./client";

export const InstructorsAPI = {
  list: () => api.get<Instructor[]>("/instructors/"),
};
