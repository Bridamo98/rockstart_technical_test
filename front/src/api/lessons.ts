import type { Lesson, LessonCreateDTO, LessonUpdateDTO } from "../types";
import { api } from "./client";

export const LessonsAPI = {
  list: (courseId: number) =>
    api.get<Lesson[]>(`/courses/${courseId}/lessons/`),
  get: (courseId: number, id: number) =>
    api.get<Lesson>(`/courses/${courseId}/lessons/${id}`),
  create: (courseId: number, payload: LessonCreateDTO) =>
    api.post<Lesson>(`/courses/${courseId}/lessons/`, payload),
  update: (courseId: number, id: number, payload: LessonUpdateDTO) =>
    api.put<Lesson>(`/courses/${courseId}/lessons/${id}`, payload),
  remove: (courseId: number, id: number) =>
    api.delete(`/courses/${courseId}/lessons/${id}`),
};
