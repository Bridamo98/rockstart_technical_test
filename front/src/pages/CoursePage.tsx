import { useEffect, useState } from "react";
import { useNavigate, useParams } from "react-router-dom";
import { CoursesAPI } from "../api/courses";
import { LessonsAPI } from "../api/lessons";
import ConfirmModal from "../components/ConfirmModal";
import CourseForm from "../components/CourseForm";
import LessonEditModal from "../components/LessonEditModal";
import LessonForm from "../components/LessonForm";
import LessonItem from "../components/LessonItem";
import type {
  Course,
  Lesson,
  LessonCreateDTO,
  LessonUpdateDTO,
} from "../types";

export default function CoursePage() {
  const { id } = useParams();
  const navigate = useNavigate();
  const courseId = Number(id);
  const [course, setCourse] = useState<Course | null>(null);
  const [lessons, setLessons] = useState<Lesson[]>([]);
  const [editMode, setEditMode] = useState(false);
  const [showLessonForm, setShowLessonForm] = useState(false);
  const [editingLesson, setEditingLesson] = useState<Lesson | null>(null);
  const [confirmDeleteCourse, setConfirmDeleteCourse] = useState(false);
  const [lessonToDelete, setLessonToDelete] = useState<Lesson | null>(null);
  const [pendingCourseChanges, setPendingCourseChanges] =
    useState<Partial<Course> | null>(null);

  const saveLesson = (changes: LessonUpdateDTO) =>
    editingLesson &&
    LessonsAPI.update(courseId, editingLesson.id, changes).then(() => {
      setEditingLesson(null);
      fetchAll();
    });

  const fetchAll = () => {
    CoursesAPI.get(courseId).then((r) => setCourse(r.data));
    LessonsAPI.list(courseId).then((r) => setLessons(r.data));
  };

  useEffect(() => {
    fetchAll();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const askUpdateCourse = (data: Partial<Course>) =>
    setPendingCourseChanges(data);

  const updateCourse = (data: Partial<Course>) =>
    CoursesAPI.update(courseId, data)
      .then(fetchAll)
      .then(() => setEditMode(false));

  const deleteCourse = () =>
    CoursesAPI.remove(courseId).then(() => navigate("/courses"));

  const addLesson = (formData: { title?: string; video_url?: string }) => {
    if (!formData.title || !formData.video_url) {
      return;
    }
    const data: LessonCreateDTO = {
      title: formData.title,
      video_url: formData.video_url,
    };
    return LessonsAPI.create(courseId, data)
      .then(fetchAll)
      .then(() => setShowLessonForm(false));
  };

  const delLesson = (lessonId: number) =>
    LessonsAPI.remove(courseId, lessonId).then(fetchAll);

  return (
    <div className="container mx-auto p-6 space-y-6">
      <button
        onClick={() => navigate("/courses")}
        className="btn btn-outline mb-4"
      >
        ‚Üê Back to Courses
      </button>
      {course && (
        <div className="flex items-center justify-between">
          <div>
            <h1 className="text-2xl font-bold">{course.title}</h1>
          </div>
          <div className="space-x-2">
            {!editMode && (
              <>
                <button
                  onClick={() => setEditMode(true)}
                  className="btn btn-secondary"
                >
                  Edit
                </button>
                <button
                  onClick={() => setConfirmDeleteCourse(true)}
                  className="btn btn-error"
                >
                  Delete
                </button>
              </>
            )}
            {editMode && (
              <button
                onClick={() => setEditMode(false)}
                className="btn btn-outline"
              >
                Cancel
              </button>
            )}
          </div>
        </div>
      )}

      {course && !editMode && <p>{course.course_desc}</p>}

      {editMode && course && (
        <CourseForm
          defaultValues={course as Course}
          onSubmit={askUpdateCourse}
          submitLabel="Update"
        />
      )}

      <hr />

      <div className="flex items-center justify-between">
        <h2 className="text-xl font-semibold">Lessons</h2>
        <button
          onClick={() => setShowLessonForm(!showLessonForm)}
          className="btn btn-primary"
        >
          {showLessonForm ? "Cancel" : "Add lesson"}
        </button>
      </div>

      {showLessonForm && <LessonForm onSubmit={addLesson} />}

      <div className="grid md:grid-cols-2 gap-4">
        {lessons.map((l) => (
          <LessonItem
            key={l.id}
            lesson={l}
            onEdit={() => setEditingLesson(l)}
            onDelete={() => setLessonToDelete(l)}
          />
        ))}
      </div>
      <LessonEditModal
        open={!!editingLesson}
        lesson={editingLesson}
        onClose={() => setEditingLesson(null)}
        onSave={saveLesson}
      />
      <ConfirmModal
        open={confirmDeleteCourse}
        title="Delete course?"
        message="This action cannot be undone."
        confirmLabel="Delete"
        onCancel={() => setConfirmDeleteCourse(false)}
        onConfirm={() => {
          setConfirmDeleteCourse(false);
          deleteCourse();
        }}
      />
      <ConfirmModal
        open={!!pendingCourseChanges}
        title="Save changes?"
        message="Modify course details?"
        confirmLabel="Save"
        onCancel={() => setPendingCourseChanges(null)}
        onConfirm={() => {
          updateCourse(pendingCourseChanges!);
          setPendingCourseChanges(null);
        }}
      />
      <ConfirmModal
        open={!!lessonToDelete}
        title="Delete lesson?"
        message={`"${lessonToDelete?.title}" will be removed.`}
        confirmLabel="Delete"
        onCancel={() => setLessonToDelete(null)}
        onConfirm={() => {
          delLesson(lessonToDelete!.id);
          setLessonToDelete(null);
        }}
      />
    </div>
  );
}
