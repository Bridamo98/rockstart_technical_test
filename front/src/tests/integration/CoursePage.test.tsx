import { screen, waitFor, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { beforeEach, expect, test } from "vitest";
import { CoursesAPI } from "../../api/courses";
import CoursePage from "../../pages/CoursePage";
import { resetMockState } from "../mocks/handlers";
import { renderCourseRoute } from "../testUtils";

async function makeCourse() {
  const { data } = await CoursesAPI.create({
    title: "Initial",
    course_desc: "desc",
    instructor_id: 1,
  });
  return data.id;
}

beforeEach(resetMockState);

test("muestra info y permite actualizar título", async () => {
  const id = await makeCourse();
  const { asFragment } = renderCourseRoute(id, <CoursePage />);
  const user = userEvent.setup();

  await screen.findByText("Initial");

  await user.click(screen.getByRole("button", { name: /edit/i }));
  const titleInput = screen.getByPlaceholderText(/title/i);
  await user.clear(titleInput);
  await user.type(titleInput, "Updated!");
  await user.click(screen.getByRole("button", { name: /update/i }));

  await user.click(screen.getByRole("button", { name: /^save$/i }));

  await waitFor(() => expect(screen.getByText("Updated!")).toBeInTheDocument());

  expect(asFragment()).toMatchSnapshot("CoursePage-after-update");
});

test("agrega y elimina una lección", async () => {
  const id = await makeCourse();
  const { asFragment } = renderCourseRoute(id, <CoursePage />);
  const user = userEvent.setup();

  expect(asFragment()).toMatchSnapshot("CoursePage-before-add-lesson");

  await user.click(screen.getByRole("button", { name: /add lesson/i }));
  await user.type(screen.getByPlaceholderText(/lesson title/i), "L1");
  await user.type(
    screen.getByPlaceholderText(/youtube url/i),
    "https://www.youtube.com/watch?v=ABCDEFGHIJK"
  );
  await user.click(screen.getByRole("button", { name: /^save$/i }));

  await screen.findByText("L1");

  expect(asFragment()).toMatchSnapshot("CoursePage-after-add-lesson");

  const lessonHeading = await screen.findByText("L1", { selector: "h4" });
  const card = lessonHeading.closest("div");
  const deleteLessonBtn = within(card!).getByRole("button", {
    name: /^delete$/i,
  });

  await user.click(deleteLessonBtn);

  await screen.findByText(/delete lesson\?/i);

  expect(document.body).toMatchSnapshot("CoursePage-before-click-delete");

  const deleteButtons = screen.getAllByRole("button", { name: /^delete$/i });
  await user.click(deleteButtons[2]);
  await waitFor(() =>
    expect(screen.queryByText(/delete lesson\?/i)).not.toBeInTheDocument()
  );
  expect(asFragment()).toMatchSnapshot("CoursePage-after-click-delete");
});
