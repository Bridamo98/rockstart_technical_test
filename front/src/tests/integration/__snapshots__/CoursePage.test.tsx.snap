// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`agrega y elimina una lección > CoursePage-after-add-lesson 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto p-6 space-y-6"
  >
    <button
      class="btn btn-outline mb-4"
    >
      ← Back to Courses
    </button>
    <div
      class="flex items-center justify-between"
    >
      <div>
        <h1
          class="text-2xl font-bold"
        >
          Initial
        </h1>
      </div>
      <div
        class="space-x-2"
      >
        <button
          class="btn btn-secondary"
        >
          Edit
        </button>
        <button
          class="btn btn-error"
        >
          Delete
        </button>
      </div>
    </div>
    <p>
      desc
    </p>
    <hr />
    <div
      class="flex items-center justify-between"
    >
      <h2
        class="text-xl font-semibold"
      >
        Lessons
      </h2>
      <button
        class="btn btn-primary"
      >
        Add lesson
      </button>
    </div>
    <div
      class="grid md:grid-cols-2 gap-4"
    >
      <div
        class="bg-white p-4 rounded shadow space-y-2"
      >
        <h4
          class="font-bold"
        >
          L1
        </h4>
        <div
          class="aspect-video"
        >
          <iframe
            allowfullscreen=""
            class="w-full h-full"
            src="https://www.youtube.com/embed/ABCDEFGHIJK"
            title="video_ABCDEFGHIJK"
          />
        </div>
        <div
          class="flex gap-2"
        >
          <button
            class="btn btn-sm btn-secondary"
          >
            Edit
          </button>
          <button
            class="btn btn-sm btn-error"
          >
            Delete
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`agrega y elimina una lección > CoursePage-after-click-delete 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto p-6 space-y-6"
  >
    <button
      class="btn btn-outline mb-4"
    >
      ← Back to Courses
    </button>
    <div
      class="flex items-center justify-between"
    >
      <div>
        <h1
          class="text-2xl font-bold"
        >
          Initial
        </h1>
      </div>
      <div
        class="space-x-2"
      >
        <button
          class="btn btn-secondary"
        >
          Edit
        </button>
        <button
          class="btn btn-error"
        >
          Delete
        </button>
      </div>
    </div>
    <p>
      desc
    </p>
    <hr />
    <div
      class="flex items-center justify-between"
    >
      <h2
        class="text-xl font-semibold"
      >
        Lessons
      </h2>
      <button
        class="btn btn-primary"
      >
        Add lesson
      </button>
    </div>
    <div
      class="grid md:grid-cols-2 gap-4"
    />
  </div>
</DocumentFragment>
`;

exports[`agrega y elimina una lección > CoursePage-before-add-lesson 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto p-6 space-y-6"
  >
    <button
      class="btn btn-outline mb-4"
    >
      ← Back to Courses
    </button>
    <hr />
    <div
      class="flex items-center justify-between"
    >
      <h2
        class="text-xl font-semibold"
      >
        Lessons
      </h2>
      <button
        class="btn btn-primary"
      >
        Add lesson
      </button>
    </div>
    <div
      class="grid md:grid-cols-2 gap-4"
    />
  </div>
</DocumentFragment>
`;

exports[`agrega y elimina una lección > CoursePage-before-click-delete 1`] = `
<body>
  <div>
    <div
      class="container mx-auto p-6 space-y-6"
    >
      <button
        class="btn btn-outline mb-4"
      >
        ← Back to Courses
      </button>
      <div
        class="flex items-center justify-between"
      >
        <div>
          <h1
            class="text-2xl font-bold"
          >
            Initial
          </h1>
        </div>
        <div
          class="space-x-2"
        >
          <button
            class="btn btn-secondary"
          >
            Edit
          </button>
          <button
            class="btn btn-error"
          >
            Delete
          </button>
        </div>
      </div>
      <p>
        desc
      </p>
      <hr />
      <div
        class="flex items-center justify-between"
      >
        <h2
          class="text-xl font-semibold"
        >
          Lessons
        </h2>
        <button
          class="btn btn-primary"
        >
          Add lesson
        </button>
      </div>
      <div
        class="grid md:grid-cols-2 gap-4"
      >
        <div
          class="bg-white p-4 rounded shadow space-y-2"
        >
          <h4
            class="font-bold"
          >
            L1
          </h4>
          <div
            class="aspect-video"
          >
            <iframe
              allowfullscreen=""
              class="w-full h-full"
              src="https://www.youtube.com/embed/ABCDEFGHIJK"
              title="video_ABCDEFGHIJK"
            />
          </div>
          <div
            class="flex gap-2"
          >
            <button
              class="btn btn-sm btn-secondary"
            >
              Edit
            </button>
            <button
              class="btn btn-sm btn-error"
            >
              Delete
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="fixed inset-0 z-50 flex items-center justify-center bg-black/50"
  >
    <div
      class="bg-white rounded shadow p-6 w-full max-w-lg"
    >
      <h3
        class="text-lg font-semibold mb-3"
      >
        Delete lesson?
      </h3>
      <p
        class="mb-6"
      >
        "L1" will be removed.
      </p>
      <div
        class="flex justify-end gap-2"
      >
        <button
          class="btn btn-secondary"
        >
          Cancel
        </button>
        <button
          class="btn btn-error"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`muestra info y permite actualizar título > CoursePage-after-update 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto p-6 space-y-6"
  >
    <button
      class="btn btn-outline mb-4"
    >
      ← Back to Courses
    </button>
    <div
      class="flex items-center justify-between"
    >
      <div>
        <h1
          class="text-2xl font-bold"
        >
          Updated!
        </h1>
      </div>
      <div
        class="space-x-2"
      >
        <button
          class="btn btn-secondary"
        >
          Edit
        </button>
        <button
          class="btn btn-error"
        >
          Delete
        </button>
      </div>
    </div>
    <p>
      desc
    </p>
    <hr />
    <div
      class="flex items-center justify-between"
    >
      <h2
        class="text-xl font-semibold"
      >
        Lessons
      </h2>
      <button
        class="btn btn-primary"
      >
        Add lesson
      </button>
    </div>
    <div
      class="grid md:grid-cols-2 gap-4"
    />
  </div>
</DocumentFragment>
`;
